game/src/game/services/GameStateService.cpp

```diff
@@ -1195,73 +1260,113 @@ std::unordered_set<std::string> GameStateService::GetPlayerStaticNames()
     return names;
 }
 
-void GameStateService::UpdatePowerupInfo()
+std::ofstream& operator<<(std::ofstream& file, const glm::vec3& vec)
 {
-    // assigning powerup info here.
+    file << vec.x << ' ' << vec.y << ' ' << vec.z;
+    return file;
+}
 
-    // extract the information from obj file about every powerup location.
-    Checkpoints checkpoints;
-    auto locations = checkpoints.GetCheckpoints();
+std::vector<std::pair<glm::vec3, std::string>>
+GameStateService::ReadCheckpointsFromJsonFile()
+{
+    std::vector<std::pair<glm::vec3, std::string>> data;
 
-    // locations.second = orientation, we do not care about that, ignore that
-    // for now
+    std::ifstream file("resources/powerup/PowerupSpawnPoint.jsonc");
+    if (!file.is_open())
+    {
+        debug::LogError("Powerup Spawn not happening");
+    }
 
-    for (const auto& l : locations)
+    std::string jsonStr((std::istreambuf_iterator<char>(file)),
+                        std::istreambuf_iterator<char>());
+    rapidjson::Document document;
+    document.Parse(jsonStr.c_str());
+
+    if (document.IsArray())
     {
-        // adding a random statement which will select to spawn ammo type or
-        // powerup at the lcoation randomly generating what kind of powerup
-        uint8_t selector = rand() % 2 + 1;
-        if (selector == 1)
+        for (rapidjson::SizeType i = 0; i < document.Size(); i++)
         {
-            // means ammo should spawn
-            uint8_t random_ammo = rand() % 5 + 1;
-            switch (random_ammo)
+            const rapidjson::Value& object = document[i];
+
+            if (object.HasMember("Location") && object.HasMember("Pickup_Type"))
             {
-                case 1:
-                    ammo_info_.push_back({AmmoPickupType::kBuckshot, l.first});
-                    break;
-                case 2:
-                    ammo_info_.push_back(
-                        {AmmoPickupType::kDoubleDamage, l.first});
-                    break;
-                case 3:
-                    ammo_info_.push_back(
-                        {AmmoPickupType::kExploadingBullet, l.first});
-                    break;
-                case 4:
-                    ammo_info_.push_back(
-                        {AmmoPickupType::kIncreaseFireRate, l.first});
-                    break;
-                case 5:
-                    ammo_info_.push_back(
-                        {AmmoPickupType::kVampireBullet, l.first});
-                    break;
+                const rapidjson::Value& location = object["Location"];
+                const rapidjson::Value& pickup_type = object["Pickup_Type"];
```

Comment:
There's some existing utility func for getting a string from JSON to make it easier:
![image](https://user-images.githubusercontent.com/41604234/232122216-eb8e6a3f-086c-45e3-b612-7f5b7ceadbc8.png)
[Code example from AssetBundle.cpp like ~36](https://github.com/raduschirliu/cpsc585/blob/6167c4eeb9d8e1cf45faec3e243d4f178d6973d9/game/src/engine/asset/AssetBundle.cpp#L44)
and the util func is in `#include "engine/core/json/deserialize_utils.h"`